src/main/scala/com/fulcrumgenomics/bam/BaseCounts.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.FgBioDef._
+import htsjdk.samtools.util.SamLocusIterator
+import htsjdk.samtools.util.SamLocusIterator.LocusInfo
+import scala.collection.JavaConversions.iterableAsScalaIterable
+
+object BaseCounts {
+  /**
+    * Generates a BaseCounts object from a collection of LocusInfo object
+    * representing a pileup.  Ensures that each template is only counted once.
+    */
+  def apply(locus: LocusInfo): BaseCounts = apply(locus.getRecordAndOffsets)
+
+  /**
+    * Generates a BaseCounts object from a collection of RecordAndOffest objects
+    * representing a pileup.  Ensures that each template is only counted once.
+    */
+  def apply(recs: Traversable[SamLocusIterator.RecordAndOffset]): BaseCounts = {
+    val countsByBase = recs.groupBy(r => Character.toUpperCase(r.getReadBase.toChar))
+      .map { case (ch, rs) => (ch, rs.map(_.getRecord.getReadName).toSeq.distinct.size) }
```

Comment:
Yes, this is a way to only count a given template once to avoid double-counting over-lapping reads.  The filtering by base quality and mapping quality is done internally by the `SamLocusIterator`.